contracts/artifacts/DelegatedOptIn.approval.teal

```diff
@@ -0,0 +1,165 @@
+#pragma version 9
+
+txn ApplicationID
+int 0
+>
+int 6
+*
+txn OnCompletion
++
+switch create_NoOp NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED NOT_IMPLEMENTED call_NoOp
+
+NOT_IMPLEMENTED:
+	err
+
+abi_route_setSignature:
+	byte 0x
+	txn GroupIndex
+	int 1
+	-
+	txna ApplicationArgs 1
+	callsub setSignature
+	int 1
+	return
+
+setSignature:
+	proto 3 0
+
+	// Record MBR before box_put to later determine the MBR delta
+	// ./contracts/delegated_optin_app.algo.ts:22
+	// preMBR = this.app.address.minBalance
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	acct_params_get AcctMinBalance
+	assert
+	frame_bury -3 // preMBR: uint64
+
+	// ./contracts/delegated_optin_app.algo.ts:23
+	// this.signatures.set(this.txn.sender, sig)
+	txn Sender
+	frame_dig -1 // sig: byte[64]
+	box_put
+
+	// Verify box MBR payment
+	// ./contracts/delegated_optin_app.algo.ts:26
+	// assert(boxMBRPayment.receiver === this.app.address)
+	frame_dig -2 // boxMBRPayment: pay
+	gtxns Receiver
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	==
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:27
+	// assert(boxMBRPayment.amount >= this.app.address.minBalance - preMBR)
+	frame_dig -2 // boxMBRPayment: pay
+	gtxns Amount
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	acct_params_get AcctMinBalance
+	assert
+	frame_dig -3 // preMBR: uint64
+	-
+	>=
+	assert
+	retsub
+
+abi_route_delegatedOptIn:
+	// no dupn needed
+	txn GroupIndex
+	int 1
+	-
+	txn GroupIndex
+	int 2
+	-
+	callsub delegatedOptIn
+	int 1
+	return
+
+delegatedOptIn:
+	proto 2 0
+
+	// Verify that the signature is present
+	// ./contracts/delegated_optin_app.algo.ts:39
+	// assert(this.signatures.exists(optIn.sender))
+	frame_dig -2 // optIn: axfer
+	gtxns Sender
+	box_len
+	swap
+	pop
+	assert
+	retsub
```

Comment:
This function never checks the pay?

Reply:
I believe he's counting on the signed logicsig to check.

I think they both should. It's just easier to understand if a reader of each program reason without seeing the other.  Of course, you can't reason about *all* the security properties without seeing them in combination. 

In particular, reading the logicsig alone can't guarantee that the app handles revocations properly, but it can protect itself from anything except compensated opt-ins (spam?).  Meanwhile, the app could be written in a way that if I know my logicsig requires the app's existence, I'm safe, regardless of what else the logicsig I'm signing does. I could imagine, for example, that if this was at the top:
```
// Verify called atomically with master app
AppCall
gtxns ApplicationID
int TMPL_DELEGATED_OPTIN_APP_ID
==
assert

// Verify the correct method is being called
AppCall
gtxnsa ApplicationArgs 0
method "delegatedOptIn(pay,axfer)void"
==
assert
```
Then there could be wallets that say "You're safe, you're being asked to signed a logicsig for an app/method that is on my allowlist of known trustworthy apps."

I know this is not strictly necessary, but we should be making auditing trivial and mistakes as impossible as we can.

I even think the app should check things the logicsig checks, like Fee=0, no re-key. Both should be as clearly harmless as can be.

Reply:
Yeah I know this was originally discussed. I thought the reasoning for having checks in the app was because some checks needed to be in the app and some needed to be in the lsig, but now all checks can be in the lsig so it's less confusing.

I have no problem with adding checks to the app though so will add in the next update

Reply:
I have moved app verification to the top of the TEAL in 75ceab28e1f95ec754bfe0b1452e27ace2806206 and added transaction verification to the app in 4701fd7d4b0204e167a92182d36262b8d48bcb4a.